design-proposals/vhostuser/vhostuser-interface.md

```diff
@@ -0,0 +1,143 @@
+
+## Motivation
+- Any kubevirt VM uses linux kernel interfaces to connect to its host and to reach out externally or to communicate within the cluster. The current kubevirt VMs can only support the kernel data path for which the traffic needs to be through the kernel of the node/compute hosting the VM.
+
+- As a result, when we have a host with DPDK dataplane running in Userspace, the traffic from kubevirt VM will go through the kernel and reach the dataplane running in userspace. This longer path slows down the traffic between the application and the dataplane when both are running in userspace.
+
+- In order to improve the traffic performance and make use of the dpdk, we need to use a different path that can by-pass the kernel and reach the dataplane from the user space directly. This so called fast path will run within the user space and will allow fast packet processing.
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a multus userspace CNI is being used, for instance userspace cni from intel.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane and userspace cni with multus support.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. An EmptyDIR volume (shared-dir) and DownwardAPI (podinfo) volume are mounted on the virtlauncher pod to have the support for new interface.
+- These mounts will allow the virt-launcher pod to create a VM with an additional interface which can be reached using the vhostSocket mentioned in the DownwardAPI.
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. **Once the VM spec virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) "/var/lib/cni/usrcni/" is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane. This socket file acts as an UNIX socket between the VM interface and the dataplane running in the UserSpace.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
```

Comment:
If the `EmptyDir` mounts the actual socket to a specific path, why do we need the DownwardAPI ?
What information needs to arrive to the virt-launcher filesystem and why it cannot be passed through a field in the VMI (e.g. the in the interface details).

Reply:
@EdDev , Also the CNI/KubeManager will be creating the socket files while creating the virtio interfaces. So, to have the consistency, we expect that the CNI uses the socket name reperesenting the poduid+intfname+id. This info will not be available from the VM YAML spec.

DownwardAPI will help us maintain consistency as the CNI creates the sockets in a custom path, which will be shared to the launcher pod. This socket name will be passed to qemu.

Reply:
> @EdDev , Also the CNI/KubeManager will be creating the socket files while creating the virtio interfaces. So, to have the consistency, we expect that the CNI uses the socket name reperesenting the poduid+intfname+id. This info will not be available from the VM YAML spec.
> 
> DownwardAPI will help us maintain consistency as the CNI creates the sockets in a custom path, which will be shared to the launcher pod. This socket name will be passed to qemu.

@nvinnakota10, if socket name is built after `poduid+intfname+id`, live migration of VM will fail because a new pod  will be created where the running VM will be transferred to, and qemu will be started with the same parameters including the path/socketname.
But obviously `poduid` changed and the CNI created a new socket named after `newpodui+intfname+id`, and uses also this name to configure the vswitch port. qemu will then hang waiting the socket `poduid+intfname+id` to be opened by the client (vswitch).
We did experience this issue with userspace CNI, and had to explicitly specify socket name in the userspace CNI config in order to have a successful VM live migration:

```yaml
        "cniVersion": "0.3.1",
        "type": "userspace",
        "name": "userspace-vpp-250-1",
        "KubeConfig": "/etc/cni/net.d/multus.d/multus.kubeconfig",
        "SharedDir": "/var/lib/cni/usrcni/",
        "logFile": "/var/log/userspace-vpp-250-1.log",
        "logLevel": "debug",
        "host": {
                "engine": "vpp",
                "iftype": "vhostuser",
                "netType": "bridge",
                "vhost": {
                        "mode": "client",
                        "socketfile": "net1"
                },
                "bridge": {
                        "bridgeName": "250"
                }
        }
```

The socket name is `poduid+intfname+id` is the default one in userspace CNI. Maybe it worth changing this default naming in the CNI (standard interface?) or change the qemu parameters regarding the vhost-user-net socketname when live migrate a VM?

Reply:
@TReZ42 , Yes, we will be have to look at the Migration part, as the new pod created might still have the same socket name, but the downwardAPI might create a new socket config based on the podUID. This might cause the VM to fail waiting for the socket.

I will check the case with multi interface as well. I am not sure if the interface supports live migration. Though it is a secondary interface, it can fail live migration to fail if not handled appropriately, as the socket names might change, as PodUId changes and the VM should have the appropriate socket name.

Reply:
@TReZ42 , One possible way is to hashing the combination of name+namespace of the VM created. But for this, can you please confirm if we can Live Migrate the VMs using VirtualMachineInstanceReplicaSet kind. How do we handle the names of the VM when VirtualMachineInstanceReplicaSet kind is used to create the VMs.

Reply:
@nvinnakota10, VMI names are created after the template `name` metadata if it exists, or VirtualMachineInstanceReplicaSet name if not, appended with an id: `testvmi5wfxp`, `testvmibjkpz`.

We can also think about VirtualMachinePool: the VMs and VMIs names are created after VirtualMachinePool name, appended with -0, -1, etc. : `test-vmp-0`, `test-vmp-1`.

We can Live Migrate each of the VMI created by VirtualMachineInstanceReplicaSet or VirtualMachinePool, but not directly the VirtualMachineInstanceReplicaSet or the VirtualMachinePool.

VMI name does not change after live migration, so use a hash of `vmi+namespace+intfname` as socket name could be OK. But userspace CNI must also be adapted to use socket names according to this naming convention...

Don't you think it is possible to let the CNI defines the socket name, and for live migration patch the domain of the target VMI with the new socket name ?

Reply:
@TReZ42 , Yes. This part needs to be handled by the CNI. For the changes I am planning to upstream for vHostUser interface, Kubevirt need no information on how the socket name is derived. But the CNI needs to make sure that the socket name needs to be consistent, if they need Live Migration.

Also, if we create VMs using replicasets, the hash appended at the end of the VMI name is immutable. So, even if we Live Migrate these VMs they still have the same name in the VMI object, which shouldn't affect the Socket name consistency.

I was asking the question to make sure if the approach make sense, or do I need to look at other possibilities for having CNI to maintain consistent names.

Reply:
@nvinnakota10, if I understand correctly, you want the CNI to derive the socket name from the VM name (VMI, VMIRS, VMPool) ? But is this info available to the CNI in Downward API? Perhaps in annotations?

It's not possible to change the target domain with updated socket name, generated by the CNI as usual, the way you are doing in `createDomainInterfaces` with the vhostuser case?

Reply:
Yes, I am using the createDomainInterfaces. As soon as the CNI creates the socket name and interface name, it will store the annotations in the path mounted for DownwardAPI for the launcher pod.

The createDomainInterfaces will retrieve this information to create the socket using qemu args.